### PR TITLE
fix(ds): Support opening data sets in binary mode

### LINKED ISSUE #2696: Data set members may fail to open in binary mode
Data set members cannot always be opened in binary mode depending on the contents.

The following member can be opened in binary mode (as long as there is no trailing line) because it is exactly 80 bytes long:
```
Hello World                                                                    $
```

Attempting to open the following member in binary mode results in an error:
```
Hello World $
```

Here are the attributes of the PDS:

### PR DESCRIPTION
**Pull request by t1m0thyj**
## Proposed changes

Fixes #2696 by adding a prompt to download data sets in binary mode similar to the existing behavior for USS files.

## Release Notes

Milestone: 2.16.0

Changelog: Fixed issue where data sets or members containing binary content cannot be opened

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [ ] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ ] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [ ] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [x] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [ ] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

### CODE REVIEW

File: packages/zowe-explorer/src/dataset/DatasetTree.ts [hunk 1]
```diff
@@ -1522,8 +1523,8 @@ export class DatasetTree extends ZoweTreeProvider implements IZoweTree<IZoweData
         }
     }
 
-    public async openWithEncoding(node: IZoweDatasetTreeNode): Promise<void> {
-        const encoding = await promptForEncoding(node);
+    public async openWithEncoding(node: IZoweDatasetTreeNode, encoding?: ZosEncoding): Promise<void> {
+        encoding = encoding ?? (await promptForEncoding(node));
```
Comment (zFernand0) [inline packages/zowe-explorer/src/dataset/DatasetTree.ts:1527]: have you considered using `||=`? (e.g. `encoding ||= await prompt...`) 😋

File: packages/zowe-explorer/__tests__/__unit__/uss/ZoweUSSNode.unit.test.ts [hunk 2]
```diff
@@ -1386,84 +1386,6 @@ describe("ZoweUSSNode Unit Tests - Function node.openedDocumentInstance()", () =
     });
 });
 
-describe("ZoweUSSNode Unit Tests - Function node.initializeFileOpening()", () => {
```
Comment (zFernand0) [inline packages/zowe-explorer/__tests__/__unit__/uss/ZoweUSSNode.unit.test.ts:1389]: thanks for cleaning up these tests after exporting/moving the function outside the node

File: packages/zowe-explorer/src/shared/utils.ts [hunk 3]
```diff
@@ -544,3 +544,42 @@ export async function promptForEncoding(node: IZoweDatasetTreeNode | IZoweUSSTre
...
+                localize("zowe.shared.utils.initializeFileOpening.message", "Failed to open file as text. Re-download file as binary?"),
+                { items: [yesResponse, noResponse] }
+            );
+
+            if (response === yesResponse) {
+                const command = isZoweDatasetTreeNode(node) ? "zowe.ds.openWithEncoding" : "zowe.uss.openWithEncoding";
+                await vscode.commands.executeCommand(command, node, { kind: "binary" });
+            }
+        } else {
+            await Gui.showTextDocument(document, { preview: previewFile });
```
Comment (zFernand0) [inline packages/zowe-explorer/src/shared/utils.ts:579]: thanks for simplifying some of the logic too 😋

**Review by zFernand0 — COMMENTED** [review]
Changes LGTM! 😋 
Left a small comment but nothing that needs to be changed.
Will approve after doing some more testing 😋

**Review by traeok — DISMISSED** [review]
LGTM, thanks Timothy! I tested this with a "hello world" binary that I uploaded - in 2.15.4 I get an error message when trying to open it, but with this fix, I can open and save the PDS member as expected.

I noticed that you can right-click -> "Open with Encoding" on a member of a binary PDS and select a different encoding, but it fails to open and will prompt the user again to re-download as binary. However, for members stored with a text encoding, I can open them as either binary or text without issue, so the current behavior seems consistent to me 👍

File: packages/zowe-explorer/src/dataset/DatasetTree.ts (hunk 1 above)
Comment (t1m0thyj) [reply packages/zowe-explorer/src/dataset/DatasetTree.ts]: Typically I think we prefer using `??` over `||` - either one would probably work but the former has more strictly defined behavior. TIL that  [`??=`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_assignment) is a thing, so I'll use that 😋

**Review by zFernand0 — APPROVED** [review]
LGTM, thanks @t1m0thyj! 😋

**Review by traeok — APPROVED** [review]

**Review by JillieBeanSim — APPROVED** [review]

**Comment by JillieBeanSim** [comment]
was getting ready to merge then seen it was against maintenance branch, @t1m0thyj I thought you mentioned rebasing to main on Tues call?

**Pull request merged**